rig-core/src/providers/gemini/completion.rs

```diff
@@ -0,0 +1,607 @@
+    pub function_declaration: FunctionDeclaration,
+    pub code_execution: Option<CodeExecution>,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct FunctionDeclaration {
+    pub name: String,
+    pub description: String,
+    pub parameters: Option<Vec<Schema>>,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ToolConfig {
+    pub schema: Option<Schema>,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CodeExecution {}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SafetySetting {
+    pub category: HarmCategory,
+    pub threshold: HarmBlockThreshold,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
+pub enum HarmBlockThreshold {
+    HarmBlockThresholdUnspecified,
+    BlockLowAndAbove,
+    BlockMediumAndAbove,
+    BlockOnlyHigh,
+    BlockNone,
+    Off,
+}
+
+// =================================================================
+// Rig Implementation Types
+// =================================================================
+
+#[derive(Clone)]
+pub struct CompletionModel {
+    client: Client,
+    pub model: String,
+}
+
+impl CompletionModel {
+    pub fn new(client: Client, model: &str) -> Self {
+        Self {
+            client,
+            model: model.to_string(),
+        }
+    }
+}
+
+impl completion::CompletionModel for CompletionModel {
+    type Response = GenerateContentResponse;
+
+    async fn completion(
+        &self,
+        mut completion_request: CompletionRequest,
+    ) -> Result<completion::CompletionResponse<GenerateContentResponse>, CompletionError> {
+        // QUESTION: Why do Anthropic/openAi implementation differ here? OpenAI adds the preamble but Anthropic does not.
+
+        let mut full_history = if let Some(preamble) = &completion_request.preamble {
+            vec![completion::Message {
+                role: "system".into(),
+                content: preamble.clone(),
+            }]
+        } else {
+            vec![]
+        };
+
+        full_history.append(&mut completion_request.chat_history);
+
+        let prompt_with_context = completion_request.prompt_with_context();
+
+        full_history.push(completion::Message {
+            role: "user".into(),
+            content: prompt_with_context,
+        });
+
+        // Handle Gemini specific parameters
+        let mut generation_config =
+            GenerationConfig::try_from(completion_request.additional_params.unwrap_or_default())?;
+
+        // Set temperature from completion_request or additional_params
+        if let Some(temp) = completion_request.temperature {
+            generation_config.temperature = Some(temp);
+        }
+
+        // Set max_tokens from completion_request or additional_params
+        if let Some(max_tokens) = completion_request.max_tokens {
+            generation_config.max_output_tokens = Some(max_tokens);
+        }
```

Comment:
Why can you set these extra parameters in two locations?

Reply:
Because it's both a Rig (CompletionRequest) optionnal param and a  Gemini GenerationConfig field.

So depending if the user fills `temperature` or `max_output_token` I need to check both locations.

Reply:
It's a little odd that it's in both, this is not how it's down in the other providers. Is this a suggestion for changing the internal DX for new providers?

Reply:
Yeah the Rig hardcoded set of parameters should be favored over whatever is in `additional_params`